### PR TITLE
Created some models

### PR DESCRIPTION
**Pull request by AragonerUA**

### CODE REVIEW

File: models/huggingface_models.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+# Does not work locally because of the size of the models.
+
+'''
+def llama_model(text: str) -> str:
+    tokenizer = AutoTokenizer.from_pretrained("daryl149/llama-2-7b-chat-hf")
+    model = AutoModel.from_pretrained("daryl149/llama-2-7b-chat-hf")
+
+
+    # tokenizer = LlamaTokenizer.from_pretrained("/output/path")
```
Comment (F47-503) [inline models/huggingface_models.py:16]: Better do not leave commented unused code like this

File: models/huggingface_models.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    # model = LlamaForCausalLM.from_pretrained("/output/path")
+    tokenized = tokenizer.encode([text])
+    output = model.generate(**tokenized)
```
Comment (F47-503) [inline models/huggingface_models.py:19]: Also move this model to GPU

File: models/huggingface_models.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    encoding = tokenizer("def print_hello_world():", return_tensors="pt").to(device)
+    encoding['decoder_input_ids'] = encoding['input_ids'].clone()
+    outputs = model.generate(**encoding, max_length=15)
+    print(tokenizer.decode(outputs[0], skip_special_tokens=True))
+'''
+
+
+def codet5_base_model(text: str):
+    tokenizer = RobertaTokenizer.from_pretrained('Salesforce/codet5-base')
+    model = T5ForConditionalGeneration.from_pretrained('Salesforce/codet5-base')
```
Comment (F47-503) [inline models/huggingface_models.py:41]: All models must be on GPU if possible

File: models/huggingface_models.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    input_ids = tokenizer(text, return_tensors="pt").input_ids
+    generated_ids = model.generate(input_ids, max_length=8)
+    print(tokenizer.decode(generated_ids[0], skip_special_tokens=True))
+
+
+def codet5_small_model(text: str):
+    tokenizer = RobertaTokenizer.from_pretrained('Salesforce/codet5-small')
+    model = T5ForConditionalGeneration.from_pretrained('Salesforce/codet5-small')
+    input_ids = tokenizer(text, return_tensors="pt").input_ids
+    generated_ids = model.generate(input_ids, max_length=8)
```
Comment (F47-503) [inline models/huggingface_models.py:51]: Also better to use the same max_length default values for generation

**Review by F47-503 — CHANGES_REQUESTED** [review]
The main request is to move models to gpu before inference

File: server/models/huggingface_models.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    print(tokenizer.decode(outputs[0]))
+
+
+def llama_model(text: str):
+    model = "codellama/CodeLlama-7b-hf"
+    tokenizer = AutoTokenizer.from_pretrained(model)
+    pipeline = transformers.pipeline(
+        "text-generation",
+        model=model,
+        torch_dtype=torch.float16,
```
Comment (bzz) [inline server/models/huggingface_models.py:73]: let's use `torch.bfloat16` but still that would be ~12Gb

I've used int4 quantisation

```
    nf4_config = BitsAndBytesConfig(
        load_in_4bit=True,
        bnb_4bit_quant_type="nf4",
        bnb_4bit_use_double_quant=True,
        bnb_4bit_compute_dtype=torch.bfloat16
    )
    model_nf4 = AutoModelForCausalLM.from_pretrained(args.model, quantization_config=nf4_config)
    tokenizer = AutoTokenizer.from_pretrained(args.model)

    pipeline = transformers.pipeline(
        "text-generation",
        model=model_nf4,
        torch_dtype=torch.bfloat16,
        tokenizer=tokenizer
    )
```

File: server/models/huggingface_models.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    for seq in sequences:
+        print(f"Result: {seq['generated_text']}")
+
+
+if __name__ == "__main__":
+    # codet5_model("123")
+    # llama_model("def print_hello_world():")
+    # codet5_base_model("def print_hello_world():")
+    # codet5_small_model("def print_hello_world():")
+    # starcoder_model("def print_hello_world():")
```
Comment (bzz) [inline server/models/huggingface_models.py:96]: let's have a

```
    parser = argparse.ArgumentParser()
    parser.add_argument("--model", type=str, default="codellama/CodeLlama-7b-hf")
```

and a separate function to apply BitsAndBytesConfig for codelama.

Then the switch to starcoder will be by changing this CLI arg.

**Comment by bzz** [comment]
ATM this can not be merged until the conflicts are resolved.

**Review by bzz — APPROVED** [review]
LGTM, thanks!

**Pull request merged**